### PR TITLE
Histogram Equalization

### PR DESCRIPTION
**Pull request by jamesmcclain**
Still Needs:
   - [x] #1665 (contains commits from that PR)
   - [x] Implicit Methods
   - [x] Unit Tests
   - [x] Resolution of questions concerning equalization of signed cell types (see pictures below)

### Singleband ###

Unequalized:
screenshot from 2016-10-09 13 02 10

Equalized (signed values being treated as unsigned by GIMP):
screenshot from 2016-10-09 13 02 17

Equalized (signed values being treated as signed by GeoServer):
screenshot from 2016-10-10 09 48 51

### Multiband ###

Original:
screenshot from 2016-10-18 14 20 19

GIMP histogram equalization (Colors -> Auto -> Equalize)
screenshot from 2016-10-18 14 20 35

Histogram equalization using this PR
screenshot from 2016-10-18 14 21 25

Original bands:
tc_ng_baghdad_iq_geo-red
tc_ng_baghdad_iq_geo-green
tc_ng_baghdad_iq_geo-blue

Equalized bands (using this PR):
out-red
out-green
out-blue

### CODE REVIEW

File: raster/src/main/scala/geotrellis/raster/equalization/HistogramEqualization.scala [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      else 0
+    }
+  }
+
+  private val cmp = new _cmp()
+
+  private def _T(cellType: CellType, cdf: Array[(Double, Double)])(x: Double): Double = {
+    val i = java.util.Arrays.binarySearch(cdf, (x, 0.0), cmp)
+    val bits = cellType.bits
+    val cdf_x =
```
Comment (lossyrob) [inline raster/src/main/scala/geotrellis/raster/equalization/HistogramEqualization.scala]: More verbose variable names or comments would make this code a lot easier to read.

File: raster/src/main/scala/geotrellis/raster/equalization/MultibandEqualizationMethods.scala [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package geotrellis.raster.equalization
+
+import geotrellis.raster.histogram.StreamingHistogram
+import geotrellis.raster.MultibandTile
+import geotrellis.util.MethodExtensions
+
+
+trait MultibandEqualizationMethods extends MethodExtensions[MultibandTile] {
+  def equalize(histograms: Array[StreamingHistogram]): MultibandTile = HistogramEqualization(self, histograms)
```
Comment (lossyrob) [inline raster/src/main/scala/geotrellis/raster/equalization/MultibandEqualizationMethods.scala]: Can this only work with streaming histograms, or our histogram type in general?

**Comment by lossyrob** [comment]
I would be very curious to see the same sort of set of screenshots for RGB imagery, for instance a before-and-after landsat w/ histogram equalization on each band.

File: raster/src/main/scala/geotrellis/raster/equalization/MultibandEqualizationMethods.scala (hunk 2 above)
Comment (jamesmcclain) [inline raster/src/main/scala/geotrellis/raster/equalization/MultibandEqualizationMethods.scala]: It currently relies on the ability to get the CDF associated with the histogram directly form it.  I could add that capability to the other histogram types, but only `StreamingHistogram` has it right now.

File: raster/src/main/scala/geotrellis/raster/equalization/HistogramEqualization.scala (hunk 1 above)
Comment (jamesmcclain) [inline raster/src/main/scala/geotrellis/raster/equalization/HistogramEqualization.scala]: Okay, will-do.

**Comment by jamesmcclain** [comment]
I have updated the PR comment with some pictures demonstrating the functionality on a multiband (non-landsat) image.

**Comment by jamesmcclain** [comment]
The restriction to `StreamingHistogram` has been removed; I believe that all comments prior to this one have been addressed.

**Comment by lossyrob** [comment]
@jamesmcclain good stuff. Can you update to master?

**Pull request merged**